index.js

```diff
@@ -1,34 +1,82 @@
 var nacl = require('tweetnacl')
+var base64 = require('base64-js')
 
-// Code from https://github.com/kaepora/miniLock
-// src/js/miniLock.js
-
-// Input: User key hash (Uint8Array), Salt (Uint8Array), callback function
-// Result: Calls scrypt which returns
-//	32 bytes of key material in a Uint8Array,
-//	which then passed to the callback.
-function getScryptKey(key, salt, callback) {
-	scrypt(key, salt, 17, 8, 32, 1000, function(keyBytes) {
-		return callback(nacl.util.decodeBase64(keyBytes))
-	}, 'base64');
+// Code inspired by:
+// https://github.com/kaepora/miniLock/blob/master/src/js/miniLock.js
+
+// Extracted from tweetnacl-util-js
+// https://github.com/dchest/tweetnacl-util-js/blob/master/nacl-util.js#L16
+function decodeUTF8 (s) {
+  var i, d = unescape(encodeURIComponent(s)), b = new Uint8Array(d.length)
+  for (i = 0; i < d.length; i++) b[i] = d.charCodeAt(i)
+  return b
 }
 
-module.exports = function session(email, password, callback) {
-  var keyHash = new BLAKE2s(nacl.box.secretKeyLength)
+// Input:
+//   key                      // User key hash (Uint8Array)
+//   salt                     // Salt (email or username) (Uint8Array)
+//   logN              = 17   // CPU/memory cost parameter (Integer 1 to 31)
+//   r                 = 8    // Block size parameter
+//   dkLen             = 64   // Length of derived key in Bytes
+//   interruptStep     = 1000 // Steps to split calculation with timeouts
+//   callback function
+//
+// Result:
+//   Returns 64 bytes of scrypt derived key material in an Array,
+//   which is then passed to the callback.
+//
+function getScryptKey (key, salt, logN, r, dkLen, interruptStep, callback) {
+  'use strict'
+
+  scrypt(key, salt, logN, r, dkLen, interruptStep, function (keyBytes) {
+    return callback(keyBytes)
+  }, null)
+}
 
-  keyHash.update(nacl.util.decodeUTF8(password))
+// Input:
+//  email       // A UTF-8 username or email
+//  password    // A UTF-8 passphrase
+//  callback    // A callback function
+//
+// Result:
+//   An object literal with keys
+//
+module.exports = function session (email, password, callback) {
+  'use strict'
 
-  var key = keyHash.digest()
-  var salt = nacl.util.decodeUTF8(email)
+  // A 32 Byte BLAKE2s hash of the password bytes
+  var keyHash = new BLAKE2s()
+  keyHash.update(decodeUTF8(password))
+  var scryptKey = keyHash.digest()
 
-  getScryptKey(key, salt, function(keyBytes) {
+  getScryptKey(scryptKey, email, 17, 8, 64, 1000, function (scryptByteArray) {
     try {
-      var keyPair = nacl.box.keyPair.fromSecretKey(keyBytes)
-      callback(null, keyPair)
-    } catch(err) {
-      callback(err)
+      var keys, seedBytesUint8Array, boxKeyPairSeed,
+        signKeyPairSeed, signKeyPair
+
+      // Convert scrypt Array of Bytes to Uint8Array
+      seedBytesUint8Array = new Uint8Array(scryptByteArray)
+
+      // First 32 Bytes of scrypt seed for encryption keys
+      // Note : first 32 Bytes are the same for dkLen 32 (old way) and 64!
```

Comment:
This fact also seems to suggest that backward-compatibility was maintained, which is fantastic. /cc @jo @grempe